src/dotnet/ToolPackage/ToolPackageFactory.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Configurer;
+using Microsoft.DotNet.Tools.Install.Tool;
+using Microsoft.Extensions.EnvironmentAbstractions;
+
+namespace Microsoft.DotNet.ToolPackage
+{
+    internal class ToolPackageFactory : IToolPackageFactory
+    {
+        private const string NameOfNestedPackageDirectory = ".pkg";
+
+        public (IToolPackageStore, IToolPackageInstaller) CreateToolPackageStoreAndInstaller(
+            DirectoryPath? nonGlobalLocation = null)
+        {
+            var toolPackageStore =
+                new ToolPackageStore(GetPackageLocation(nonGlobalLocation));
+            var toolPackageInstaller = new ToolPackageInstaller(
+                toolPackageStore,
+                new ProjectRestorer(Reporter.Output));
+
+            return (toolPackageStore, toolPackageInstaller);
+        }
+
+        private static DirectoryPath GetPackageLocation(DirectoryPath? nonGlobalLocation)
+        {
+            DirectoryPath packageLocation;
+
+            if (nonGlobalLocation.HasValue)
+            {
+                packageLocation = nonGlobalLocation.Value.WithSubDirectories(NameOfNestedPackageDirectory);
+            }
+            else
+            {
+                var cliFolderPathCalculator = new CliFolderPathCalculator();
+                packageLocation = new DirectoryPath(cliFolderPathCalculator.ToolsPackagePath);
```

Comment:
If we get buy off to re-do the package directory layout for preview2, then I think `ToolsPackagePath` and `ToolsShimPath` should maybe combined to a single `ToolsPath` (i.e. the global "tools path") and we append on the `.pkg` subdirectory when calculating the global location here.

Reply:
I don't like couple 2 things that don't need to. we moved these 2 locations several times. And we once considered, shim stored in .tools and package storage in .nupkg, who knows if we need to change it again. And I don't think there is problem having one place cliFolderPathCalculator to set the location.

Reply:
In `CliFolderPathCalculator.ToolsPackagesPath`, then, are we going to have another `".pkg"` constant we append onto `ToolShimsPath`?  I just like to minimize places we have to keep in sync.

Reply:
I think a little dup of a string structure (we could even have a static class to provide nested relation logic). It better than have them coupled in class level. Later toolpackage start to assume shim will just be a directory above. it would horrible to do the directory move again.